File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1235MaximumProfitInJobScheduling.cpp
#include "LeetCode1235MaximumProfitInJobScheduling.h"

#include <algorithm>
using namespace std;

int maxProfit(vector<vector<int>>& jobs, vector<int>& dp, int length, int index)
{
    if (dp[index] != -1)
        return dp[index];
    int nextIndex{length};
    int start{index + 1};
    int end{length - 1};
    int mid{};
    while (start <= end)
    {
        mid = (start + end) >> 1;
        if (jobs[mid][0] >= jobs[index][1])
        {
            end = mid - 1;
            nextIndex = min(nextIndex, mid);
            continue;
        }
        start = mid + 1;
    }
    if (nextIndex == length)
    {
        if (index < length - 1)
            dp[index] = max(jobs[index][2], maxProfit(jobs, dp, length, index + 1));
        else
            dp[index] = jobs[index][2];
        return dp[index];
    }
    dp[index] = max(maxProfit(jobs, dp, length, index + 1), jobs[index][2] + maxProfit(jobs, dp, length, nextIndex));
    return dp[index];
}

int LeetCode1235MaximumProfitInJobScheduling::jobScheduling(vector<int>& startTime, vector<int>& endTime,
                                                            vector<int>& profit)
{
    int length = static_cast<int>(startTime.size());
    vector<vector<int>> jobs(length, vector<int>(3, 0));
    for (int i = 0; i < length; i++)
    {
        jobs[i][0] = startTime[i];
        jobs[i][1] = endTime[i];
        jobs[i][2] = profit[i];
    }
    ranges::sort(jobs, [](const vector<int>& a, const vector<int>& b)
    {
        return a[0] < b[0];
    });
    vector<int> dp(length, -1);
    return maxProfit(jobs, dp, length, 0);
}

File: LeetCodeProblemsGo/LeetCode1235MaximumProfitInJobScheduling/LeetCode1235MaximumProfitInJobScheduling.go
package LeetCode1235MaximumProfitInJobScheduling

import (
	"sort"
)

func getMaxProfitAt(jobs [][]int, dp []int, length, index int) int {
	if dp[index] > 0 {
		return dp[index]
	}
	nextIndex := length
	begin := index + 1
	end := length - 1
	var mid int
	for begin <= end {
		mid = (begin + end) >> 1
		if jobs[mid][0] >= jobs[index][1] {
			nextIndex = min(nextIndex, mid)
			end = mid - 1
			continue
		}
		begin = mid + 1
	}
	if nextIndex == length {
		if index == length-1 {
			dp[index] = jobs[index][2]
			return dp[index]
		}
		dp[index] = max(jobs[index][2], getMaxProfitAt(jobs, dp, length, index+1))
		return dp[index]
	}
	dp[index] = max(jobs[index][2]+getMaxProfitAt(jobs, dp, length, nextIndex), getMaxProfitAt(jobs, dp, length, index+1))
	return dp[index]
}
func jobScheduling(startTime []int, endTime []int, profit []int) int {
	length := len(startTime)
	jobs := make([][]int, length)
	for i := 0; i < length; i++ {
		jobs[i] = []int{startTime[i], endTime[i], profit[i]}
	}
	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i][0] < jobs[j][0]
	})
	dp := make([]int, length)
	return getMaxProfitAt(jobs, dp, length, 0)
}
